logstash-core/src/main/java/org/logstash/Event.java

```diff
@@ -213,6 +240,62 @@ public void setField(final FieldReference field, final Object value) {
         }
     }
 
+    private boolean isTagsWithMap(final FieldReference field) {
+        return field.getPath() != null && field.getPath().length > 0 && field.getPath()[0].equals(TAGS);
+    }
+
+    private boolean isLegalTagValue(final Object value) {
+        if (value instanceof String || value instanceof RubyString || value == null) {
+            return true;
+        } else if (value instanceof List) {
+            for (Object item: (List) value) {
+                if (!(item instanceof String) && !(item instanceof RubyString)) {
+                    return false;
+                }
+            }
+            return true;
+        }
+
+        return false;
+    }
+
+
+    private void setTagsAndFailTags(final Object value) {
+        tag(TAGS_FAILURE_TAG);
+        tagFailTags(Valuefier.convert(value));
+    }
+
+    /**
+     * handle key/val pair in `tags`, add _tagsparsefailure to `tags`.
+     * rename FieldReference from `tags` to `_tags`. If existing `_tags` is a list, append value to `_tags`,
+     * eg. [tags][key] -> [_tags][list_index][key]
+     * @param field FieldReference of `tags` with map path
+     * @param value
+     */
+    private void setTagsAndFailTagsWithMap(final FieldReference field, final Object value) {
```

Comment:
Adding special-case logic to handle this one circumstance is going to set us up for surprises. We don't do similar for any of our other reserved fields, and don't handle this type of collision when a user is explicitly setting a particular field, so it is surprising to me that we do so here.

I would prefer to solve the underlying bug, so that `Event#set(FieldReference, Object)` sets the field at the specified address from `FieldReference` to the specified value from `Object`, overwriting intermediates as necessary.

This is likely a discussion for the larger team since I am _sure_ that we have users in-the-wild who are accidentally relying on this silent failure.

Reply:
> _tags can fail silently by copying a map to an existing array

I referred to the first review that we simply rename the first part of FieldReference [tags][nested][field] to [_tags][nested][field] could experience the same exception that happened in `tags` field. Cleaning up the field by `event.remove()` and then copying the value should avoid type collision. I don't think we need to handle explicitly setting an incompatible type to particular field.

Currently, a normal use case is appending every illegal `tags` value to `_tags` list. If user set `event.set('_tags', ['x','y','z'])` and `event.set("[_tags][b]", "not_ok")`, it will throw exception. To address this issue, it is better to track a new issue, not necessary to fix in this PR.

To the next reviewer: here we need to decide what preserving strategy is preferred
1. keeping all illegal values in a list form in `_tags`
2. keeping only the last illegal value in `_tags` which aligns with other reserved fields.

Both cannot handle collision when a user is explicitly setting a particular field.